gh-106318: Add string method examples

blaisep opened the pull request:
This PR reformats the examples created by @adorilson in https://github.com/python/cpython/pull/105670
I followed the discussion in a related PR https://github.com/python/cpython/pull/111743 , and then discussed it with @encukou and @CAM-Gerlach . 
We decided that the fastest way to get the content into the docs was to edit the `stdtypes.rst`  to include various suggestions and to have me [update the related PR](https://github.com/python/cpython/pull/111743#issuecomment-2125942291)
So I pulled the branch, resolved merge conflicts, edited the file and created this PR.

Mentioning others who might be interested:
@hauntsaninja , @CAM-Gerlach , @encukou 

<!-- gh-issue-number: gh-106318 -->
* Issue: gh-106318
<!-- /gh-issue-number -->


<!-- readthedocs-preview cpython-previews start -->
----
📚 Documentation preview 📚: https://cpython-previews--119445.org.readthedocs.build/en/119445/library/stdtypes.html

<!-- readthedocs-preview cpython-previews end -->

adorilson committed:
[DOC] Added snipped code in str.capitalize

adorilson committed:
[DOC] Add snippet code in str.casefold and str.lower methods

adorilson committed:
[DOC] Added snippet code in str.center method

adorilson committed:
[DOC] Added snippet code in str.count method

adorilson committed:
[DOC] fix typo

adorilson committed:
[DOC] Added snippet code in str.encode method

adorilson committed:
[DOC] Added snippet code to str.endswith method

adorilson committed:
[DOC] Added snippet code with \n to str.expandtabs method

adorilson committed:
[DOC] Added snippet code to str.index method

adorilson committed:
[DOC] Added snippet code to str.isalnum method

adorilson committed:
[DOC] Added snippet code in str.isalpha method

adorilson committed:
[DOC] Added snippet code in str.isascii method

adorilson committed:
[DOC] Added snippet code to str.isdecimal method

adorilson committed:
[Doc] Added snippet code to str.isdigit method

adorilson committed:
[Doc] Added snippet code to str.islower method

adorilson committed:
[Doc] Added snippet code to str.isnumeric method

adorilson committed:
[Doc] Added snippet code to str.isprintable method

adorilson committed:
[Doc] Added snippet code to str.isspace method

adorilson committed:
[Doc] Adde snippet code to str.istitle method

adorilson committed:
[Doc] Added snippet code to str.join method

adorilson committed:
[Doc] Added snippet code to str.ljust method

adorilson committed:
[Doc] Added snippet code to str.rjust method

adorilson committed:
[Doc] Added 'See also' in str.lstrip and str.rstrip methods from one to the another

adorilson committed:
[Doc] Added snippet code to str.maketrans method

adorilson committed:
[Doc] Added snippet code to str.translate method

adorilson committed:
[Doc] Added snippet code to str.partition method

adorilson committed:
[Doc] Added snippet code to str.replace method

adorilson committed:
[Doc] Added snippet code to str.rfind method

adorilson committed:
[Doc] Added snippet code to str.rindex method

adorilson committed:
[Doc] Added snippet code to str.rpartition method

adorilson committed:
[DOC] minor change

adorilson committed:
[Doc] Added snippet code to str.rsplit method

adorilson committed:
[Doc] Added snippet code to str.startswith method

adorilson committed:
[Doc] Added snippet code to str.swapcase method

adorilson committed:
[Doc] Added snippet code to str.upper method

adorilson committed:
Minor changes (formatting and typo)

adorilson committed:
[Doc] Making lint happy

adorilson committed:
Merge branch 'main' into str_methods_2

adorilson committed:
Update Doc/library/stdtypes.rst

Co-authored-by: Shantanu <12621235+hauntsaninja@users.noreply.github.com>

adorilson committed:
[DOC] Added some 'see also' to str methods

adorilson committed:
[DOC] Moved 'For example' to the end of the previous paragraph

adorilson committed:
[DOC] Putting a space after # in comments on str methods

adorilson committed:
[DOC] Fixed inconsistent literal block quoting

hauntsaninja committed:
fix other comments

adorilson committed:
Merge branch 'main' into str_methods_2

blaisep committed:
prepare for FF

blaisep committed:
update the worflows files

blaisep committed:
prepare for FF

blaisep committed:
Use the examples created by @adorilson

blaisep committed:
Correct typos in rST formatting

blaisep committed:
Merge pull request #1 from blaisep/pr-105670

GH-106318: resolving unrelated merge conflicts

blaisep committed:
Merge pull request #2 from blaisep/main

gh-106318: Merge pull request #1 from blaisep/pr-105670

hugovk commented on Doc/library/stdtypes.rst:
```diff
@@ -1623,7 +1623,15 @@ expression support in the :mod:`re` module).
+      >>> 'PYTHON IS AMAZING'.capitalize()
+      'Python is amazing'
+      >>> 'ǋemačka Starts With a non-english Digraph'.capitalize()
+      'ǋemačka starts with a non-english digraph'
```
```suggestion
      >>> 'ǋemačka Starts With a non-English Digraph'.capitalize()
      'ǋemačka starts with a non-English digraph'
```

hugovk commented on Doc/library/stdtypes.rst:
```diff
@@ -1696,10 +1730,27 @@ expression support in the :mod:`re` module).
+      >>> 'a tuple of suffixes'.endswith(('at', 'in'))
+      False
+
+   With optional *start*, test beginning at that position::
+
+      >>> 'a tuple of suffixes'.endswith(('at', 'es'))
+      True
```
This example doesn't use *start*, should this be part of the previous example and a new one added here?

blaisep replied:
I see,
Yes, I will make it part of the previous example

hugovk commented on Doc/library/stdtypes.rst:
```diff
@@ -1696,10 +1730,27 @@ expression support in the :mod:`re` module).
+      >>> 'a tuple of suffixes'.endswith(('at', 'es'))
+      True
+
+   With optional *end*, stop comparin at that position::
```
```suggestion
   With optional *end*, stop comparing at that position::
```

hugovk commented on Doc/library/stdtypes.rst:
```diff
@@ -1796,25 +1864,60 @@ expression support in the :mod:`re` module).
    Ideographic' of the Unicode Standard
-   <https://www.unicode.org/versions/Unicode15.1.0/ch04.pdf>`_.
-
+   <https://www.unicode.org/versions/Unicode15.1.0/ch04.pdf>`_
```
```suggestion
   <https://www.unicode.org/versions/Unicode15.1.0/ch04.pdf>`_.
```

hugovk commented on Doc/library/stdtypes.rst:
```diff
@@ -1796,25 +1864,60 @@ expression support in the :mod:`re` module).
+      >>> 'Ŧ'.isascii()
+      False
+
+   ASCII characters have code points in the range U+0000-U+007F::
```
```suggestion
   ASCII characters have code points in the range U+0000-U+007F.
```

hugovk commented on Doc/library/stdtypes.rst:
```diff
@@ -1823,20 +1926,38 @@ expression support in the :mod:`re` module).
 
+   Decimal characters are those that can be used to form numbers in base 10, e.g. U+0660, ARABIC-INDIC DIGIT ZERO::
+
+      >>> '٠١٢٣٤٥٦٧٨٩'.isdecimal()  # ARABIC-INDIC DIGIT ZERO TO NINE
```
```suggestion
   Decimal characters are those that can be used to form numbers in base 10, for example::

      >>> '٠١٢٣٤٥٦٧٨٩'.isdecimal()  # ARABIC-INDIC DIGIT ZERO to NINE
```

hugovk commented on Doc/library/stdtypes.rst:
```diff
@@ -1823,20 +1926,38 @@ expression support in the :mod:`re` module).
+
+      >>> '0123456789'.isdigit()
+      True
+      >>> '٠١٢٣٤٥٦٧٨٩'.isdigit()  # ARABIC-INDIC DIGIT ZERO TO NINE
```
```suggestion
      >>> '٠١٢٣٤٥٦٧٨٩'.isdigit()  # ARABIC-INDIC DIGIT ZERO to NINE
```

hugovk commented on Doc/library/stdtypes.rst:
```diff
@@ -1873,22 +2007,58 @@ expression support in the :mod:`re` module).
+   ::
+      >>> '0123456789'.isnumeric()
+      True
+      >>> '٠١٢٣٤٥٦٧٨٩'.isnumeric()  # ARABIC-INDIC DIGIT ZERO TO NINE
```
```suggestion
      >>> '٠١٢٣٤٥٦٧٨٩'.isnumeric()  # ARABIC-INDIC DIGIT ZERO to NINE
```

hugovk commented on Doc/library/stdtypes.rst:
```diff
@@ -2032,38 +2246,70 @@ expression support in the :mod:`re` module).
    two empty strings, followed by the string itself.
 
+   See also :meth:`partition`.
+
```
```suggestion
```

hugovk commented on Doc/library/stdtypes.rst:
```diff
@@ -2226,16 +2490,19 @@ expression support in the :mod:`re` module).
+
+   See also :meth:`upper` and :meth:`lower`.
+
+   Note that it is not necessarily true that ``s.swapcase().swapcase() == s``
```
```suggestion
   Note that it is not necessarily true that ``s.swapcase().swapcase() == s``.
```

hugovk commented on Misc/NEWS.d/next/Core and Builtins/2022-07-07-05-37-53.gh-issue-94606.hojJ54.rst:
Revert unrelated change to this file?

blaisep replied:
@hugovk , I'm not sure of the best way to handle these files because they were the result of a merge conflict resolution with the branch from the [original  PR](https://github.com/python/cpython/pull/105670)

I suppose I could cherry pick these commits into a new branch off main, but I wanted to keep the original commits with @adorilson 's work building the examples

blaisep replied:
oh, I noticed that the Checks at the end of this page show some merge conflicts. I will sit still for now and follow whatever you suggest.

hugovk replied:
Everything is squash-merged at the end, and the final commit will include both of you as co-authors so you'll both get credit.

So the easiest is probably along the lines of `git checkout origin/main [filename]`

hugovk commented on Misc/NEWS.d/next/Library/2023-10-17-16-11-03.gh-issue-52161.WBYyCJ.rst:
Revert unrelated change to this file?

hugovk commented on Modules/_xxtestfuzz/fuzz_elementtree_parsewhole_corpus/expat224_utf8_bug.xml:
Revert unrelated change to this file?

hugovk commented on Doc/library/stdtypes.rst:
```diff
@@ -1796,25 +1864,60 @@ expression support in the :mod:`re` module).
-   <https://www.unicode.org/versions/Unicode15.1.0/ch04.pdf>`_.
-
+   <https://www.unicode.org/versions/Unicode15.1.0/ch04.pdf>`_
+   See Unicode Properties section in :ref:`unicode-howto`.
```
We could add a reference label to that section and then link to it directly.

https://devguide.python.org/documentation/markup/#doc-ref-role

blaisep replied:
good suggestion, thank you!

blaisep committed:
Apply suggestions from code review

Thank you for the suggestions, @hugovk 
Next I will do the edits.

Co-authored-by: Hugo van Kemenade <1324225+hugovk@users.noreply.github.com>

blaisep committed:
Update stdtypes.rst with edits to content

- Add example for optional *start* in `str.startswith()`
- Add ref link for Unicode Properties linking to the Howto section

blaisep committed:
Update unicode.rst

Add the anchor ref for the link in the stdtypes.rst

hugovk commented on Doc/library/stdtypes.rst:
```diff
@@ -1796,24 +1864,59 @@ expression support in the :mod:`re` module).
    Ideographic' of the Unicode Standard
    <https://www.unicode.org/versions/Unicode15.1.0/ch04.pdf>`_.
-
+   See Unicode Properties section in :ref:`unicode-howto-properties`.
```
This renders like:

> See Unicode Properties section in [Unicode Properties](https://cpython-previews--119445.org.readthedocs.build/en/119445/howto/unicode.html#unicode-howto-properties).

This is enough:

```suggestion
   See :ref:`unicode-howto-properties`.
```

blaisep committed:
Update Doc/library/stdtypes.rst

Remove redundant text.

Co-authored-by: Hugo van Kemenade <1324225+hugovk@users.noreply.github.com>

blaisep committed:
Remove unrelated, conflicting files.

blaisep committed:
Merge remote-tracking branch 'upstream/main' into pr-105670

blaisep committed:
Fix trailing space and location of example.

blaisep committed:
Adjust the placement of examples more meticulously.

hugovk commented on Modules/_xxtestfuzz/fuzz_elementtree_parsewhole_corpus/expat224_utf8_bug.xml:
This change needs reverting.

hugovk commented on Doc/library/stdtypes.rst:
```diff
@@ -2920,9 +3211,8 @@ arbitrary binary data.
-   If the separator is not found, return a 3-tuple
-   containing a copy of the original sequence, followed by two empty bytes or
-   bytearray objects.
+   If the separator is not found, return a 3-tuple containing a copy of the original sequence,
+   followed by two empty bytes or bytearray objects.
 
```
In general, please avoid rewrapping, especially with no other changes, it makes it harder to review.

```suggestion
   If the separator is not found, return a 3-tuple
   containing a copy of the original sequence, followed by two empty bytes or
   bytearray objects.
```

hugovk commented on Doc/library/stdtypes.rst:
```diff
@@ -2031,38 +2255,69 @@ expression support in the :mod:`re` module).
+
+The original string is returned if *width* is less than or equal to ``len(s)``.
+
+   See also :meth:`ljust`.
```
> `py:meth reference target not found: ljust`

Hmm, I wonder why this isn't linking?

encukou replied:
The line above, ```The original string is returned if *width* is less than or equal to ``len(s)``.```, needs to be indented. Since it's not, this is no longer considered documentation for `str.rjust`.

blaisep replied:
thanks

hugovk commented on Doc/library/stdtypes.rst:
```diff
@@ -2031,38 +2255,69 @@ expression support in the :mod:`re` module).
+If the separator is not found, return a 3-tuple containing
    two empty strings, followed by the string itself.
 
+   See also :meth:`partition`.
```
This also isn't linking...

encukou commented on Doc/library/stdtypes.rst:
```diff
@@ -1622,7 +1622,15 @@ expression support in the :mod:`re` module).
+
+      >>> 'PYTHON IS AMAZING'.capitalize()
+      'Python is amazing'
+      >>> 'ǋemačka Starts With a non-English Digraph'.capitalize()
```
Starting with a lowercase digraph would make a better example:
```suggestion
      >>> 'ǌemačka Starts With a non-English Digraph'.capitalize()
```

blaisep replied:
Thank you!

encukou commented on Doc/library/stdtypes.rst:
```diff
@@ -1795,24 +1863,59 @@ expression support in the :mod:`re` module).
+
+      >>> 'a commom word'.isalpha()
+      False
+      >>> 'acommomword'.isalpha()
```
```suggestion
      >>> 'a common word'.isalpha()
      False
      >>> 'acommonword'.isalpha()
```

blaisep replied:
nice catch

encukou commented on Doc/library/stdtypes.rst:
```diff
@@ -1795,24 +1863,59 @@ expression support in the :mod:`re` module).
-   ``False`` otherwise.
+   ``False`` otherwise::
+
+      >>> 'a commom word'.isascii()
+      True
+      >>> 'acommomword'.isascii()
```
```suggestion
      >>> 'a common word'.isascii()
      True
      >>> 'acommonword'.isascii()
```

blaisep replied:
Thank you!

encukou commented on Doc/library/stdtypes.rst:
```diff
@@ -2289,10 +2574,18 @@ expression support in the :mod:`re` module).
+      >>> 'Monty Python'.upper()
+      'MONTY PYTHON'
+      >>> '𐊠'.upper().isupper()  # 'CARIAN LETTER A'
+      False
```
```suggestion
      >>> '  '.upper().isupper()  # space only
      False
```

adorilson replied:
I suggest adding the space only here, but keep the carian latter A.

blaisep commented on Doc/library/stdtypes.rst:
```diff
@@ -2031,38 +2255,69 @@ expression support in the :mod:`re` module).
+      >>> 'Python'.rjust(10, '.')
+      '....Python'
+
+The original string is returned if *width* is less than or equal to ``len(s)``.
```
```suggestion
    The original string is returned if *width* is less than or equal to ``len(s)``.
```

blaisep replied:
OK, I will indent

blaisep commented on Doc/library/stdtypes.rst:
```diff
@@ -2031,38 +2255,69 @@ expression support in the :mod:`re` module).
+      >>> "Monty Python's Flying Circus".rpartition(' ')
+      ("Monty Python's Flying", ' ', 'Circus')
+
+If the separator is not found, return a 3-tuple containing
```
```suggestion
    If the separator is not found, return a 3-tuple containing
```

blaisep replied:
another indent.

blaisep commented on Doc/library/stdtypes.rst:
```diff
@@ -2031,38 +2255,69 @@ expression support in the :mod:`re` module).
+      ("Monty Python's Flying", ' ', 'Circus')
+
+If the separator is not found, return a 3-tuple containing
    two empty strings, followed by the string itself.
```
```suggestion
    two empty strings, followed by the string itself.
```

blaisep replied:
indent

blaisep committed:
Update Doc/library/stdtypes.rst

Fix indentation

blaisep committed:
Update Doc/library/stdtypes.rst

accepted

blaisep committed:
Update Doc/library/stdtypes.rst

accepted

hugovk commented on Doc/library/stdtypes.rst:
```diff
@@ -2031,38 +2255,69 @@ expression support in the :mod:`re` module).
+      >>> 'Python'.rjust(10, '.')
+      '....Python'
+
+    The original string is returned if *width* is less than or equal to ``len(s)``.
```
I think this might fix the [Sphinx error](https://github.com/python/cpython/actions/runs/9390730454/job/25861276052?pr=119445):

```suggestion
   The original string is returned if *width* is less than or equal to ``len(s)``.
```

For some reason, reST is usually indented in multiples of 3 spaces.

blaisep committed:
WIP update some examples

blaisep committed:
Merge branch 'pr-105670' of github.com:blaisep/cpython into pr-105670

blaisep committed:
WIP Edit and change some examples

From office hours at https://www.youtube.com/live/-0SXSTkJsfg?si=QKhDzyWPVvWRLcDO&t=4267

Co-authored-by: Lisa Crispin <lisa.crispin@gmail.com>

adorilson commented on Doc/library/stdtypes.rst:
```diff
@@ -1676,26 +1676,46 @@ expression support in the :mod:`re` module).
+      >>> type(encoded_str_to_byte)
+      <class 'bytes'>
+      >>> encoded_str_to_byte
+      b'¿como estás?'
```
There is something wrong here.

blaisep replied:
Ok, thanks. I have fixed that.

adorilson commented on Doc/library/stdtypes.rst:
```diff
@@ -1784,7 +1801,7 @@ expression support in the :mod:`re` module).
 
-      >>> 'spam, spam, spam'.find('sp')
+      >>> 'spam, spam, spam'.find(',')
       0
```
Is there something wrong here?

```python
>>> 'spam, spam, spam'.find(',')
4
>>> 'spam, spam, spam'.find('sp')
0
```

blaisep replied:
Yes, these are more copy pasta errors. I will check this much more thoroughly the next time.

adorilson commented on Doc/library/stdtypes.rst:
```diff
@@ -1934,8 +1946,8 @@ expression support in the :mod:`re` module).
       True
-      >>> '²'.isdecimal(), '²'.isdigit()
-      (False, True)
+      >>> '²'.isdecimal()
+      True
```
I got

```python
>>> '²'.isdecimal()
False
```

blaisep replied:
Ok I have now fixed this too.

adorilson reviewed (changes requested):
I notice some possible errors, @blaisep

blaisep committed:
Extend the example for str.format_map

Fix doctest formatting errors

adorilson commented on Doc/library/stdtypes.rst:
```diff
@@ -1863,7 +1857,7 @@ expression support in the :mod:`re` module).
       ...
-      >>> '{name} was born in {country}'.format_map(Default(name='Guido'))
+      >>> '{name} was born in {country}'.format_map(Default(name='Guido', country='Netherlands'))
       'Guido was born in country'
```
This is supposed to be:

```
'Guido was born in Netherlands'
```

However, I think the previous example is more valuable. It shows the two Default class behaviour.
